File: M3Tools/Types/People/Customer.vb
Option Strict On

Namespace Types
	Public Class Customer
		Inherits Person
		'Private Const EmailPattern As String = "^[a-zA-Z0-9_!#$%&’*+/=?`{|}~^.-]+@[a-zA-Z0-9.-]+$"
		Private __phone As String

		<ComponentModel.Browsable(False)>
		<Text.Json.Serialization.JsonPropertyName("customerID")>
		Overrides Property Id As Integer

		<ComponentModel.Category("Contact")>
		<Text.Json.Serialization.JsonPropertyName("phoneNumber")>
		Public Property PhoneNumber As String
			Get
				Return __phone
			End Get
			Set(value As String)
				__phone = value.FormatPhone()
			End Set
		End Property

		<ComponentModel.Category("Contact")>
		<ComponentModel.TypeConverter(GetType(ComponentModel.ExpandableObjectConverter))>
		<Text.Json.Serialization.JsonPropertyName("address")>
		Public Property Address As Address

		<ComponentModel.Browsable(False)>
		<Text.Json.Serialization.JsonPropertyName("joined")>
		Public Property Joined As Date
		'	Get
		'		If __joined.Year < 1950 OrElse IsNothing(__joined) Then
		'			Return Nothing
		'		End If

		'		Return __joined
		'	End Get
		'	Set(value As Date)
		'		If value.Year < 1950 Then
		'			__joined = Nothing
		'		End If
		'	End Set
		'End Property

		Public Sub New()
			Me.New(-1, "John", "Doe", Nothing, "johndoe@domain.ext", "123-456-7890", Date.Now.ToLongDateString)
		End Sub

		Public Sub New(Optional customerID As Integer = -1, Optional firstName As String = "John", Optional lastName As String = "Doe",
					   Optional address As Address = Nothing, Optional email As String = "johndoe@domain.ext", Optional phoneNumber As String = "1234567890",
					   Optional joined As String = Nothing)
			Me.New(customerID, $"{firstName} {lastName}", address, phoneNumber, email, If(String.IsNullOrWhiteSpace(joined), Nothing, Date.Parse(joined)))
		End Sub

		Private Sub New(id As Integer, name As String, address As Address, phone As String, email As String, join As Date)
			MyBase.New(id, name, email)
			Me.PhoneNumber = phone
			Me.Address = If(address, Address.None)
			Me.Joined = join
		End Sub

		Public Overrides Function ToString() As String
			'Name (Email)
			'Street
			'City, ST ZipCode
			'Phone Number
			Return $"{Name} ({Email}){vbCrLf}
					{Address}{vbCrLf}
					{PhoneNumber}"
		End Function

		Public Function Display() As String
			'ID) Name (Email)
			'Street
			'City, ST ZipCode
			'Phone Number
			Return $"{Id}) {Name} (e: {Email} p: {PhoneNumber}){vbCrLf}{vbCrLf}{Address.Display}{vbCrLf}"
		End Function
	End Class
End Namespace

File: M3Tools/Types/People/Customer.cs
using System;

using Microsoft.VisualBasic;

using SPPBC.M3Tools.Types.Extensions;

namespace SPPBC.M3Tools.Types
{
	/// <summary>
	/// Class containing customer data
	/// </summary>
	[System.ComponentModel.TypeConverter(typeof(Converters.M3AppConvert<Customer>))]
	[Newtonsoft.Json.JsonObject]
	public class Customer : Person
	{
		// Private Const EmailPattern As String = "^[a-zA-Z0-9_!#$%&’*+/=?`{|}~^.-]+@[a-zA-Z0-9.-]+$"
		private string __phone;

		/// <summary>
		/// An empty instance of a customer
		/// </summary>
		public static new Customer None => new();

		/// <summary>
		/// 
		/// </summary>
		[Newtonsoft.Json.JsonProperty("customerID")]
		public override int Id
		{
			get => base.Id;
			set => base.Id = value;
		}

		/// <summary>
		/// The customers phone number
		/// </summary>
		[System.ComponentModel.Category("Contact")]
		[Newtonsoft.Json.JsonProperty("phoneNumber")]
		public string Phone
		{
			get => __phone;
			set => __phone = value.FormatPhone();
		}

		/// <summary>
		/// The shipping address of the customer
		/// </summary>
		[System.ComponentModel.Category("Contact")]
		[System.ComponentModel.TypeConverter(typeof(System.ComponentModel.ExpandableObjectConverter))]
		[Newtonsoft.Json.JsonProperty("address")]
		public Address Address { get; set; }

		/// <summary>
		/// The date the user was added to the database
		/// </summary>
		[Newtonsoft.Json.JsonProperty("joined")]
		public DateTime Joined { get; set; }

		/// <summary>
		/// 
		/// </summary>
		/// <param name="customerID"></param>
		/// <param name="firstName"></param>
		/// <param name="lastName"></param>
		/// <param name="address"></param>
		/// <param name="email"></param>
		/// <param name="phoneNumber"></param>
		/// <param name="joined"></param>
		[Newtonsoft.Json.JsonConstructor]
		public Customer(int customerID = -1, string firstName = "New", string lastName = "Customer", Address address = null, string phoneNumber = "1234567890", string email = "johndoe@domain.ext", string joined = null) : this(customerID, $"{firstName} {lastName}", address, phoneNumber, email, string.IsNullOrWhiteSpace(joined) ? default : DateTime.Parse(joined))
		{
		}

		private Customer(int customerID, string name, Address address, string phone, string email, DateTime @join) : base(customerID, name, email)
		{
			Phone = phone;
			Address = address ?? Address.None;
			Joined = join;
		}

		/// <summary>
		/// Returns the customer in a string format
		/// </summary>
		/// <returns></returns>
		public override string ToString() =>
			// Id,Name,Email,Address,Phone,Joined
			string.Join(Properties.Settings.Default.ObjectDelimiter, Id, Name, Email, Address.ToString(), Phone, Joined);

		/// <summary>
		/// Returns the info for the custoemr in a human readable format
		/// </summary>
		/// <returns></returns>
		public string Display() =>
			// ID) Name (Email)
			// Street
			// City, ST ZipCode
			// Phone Number
			$"{Id}) {Name} (e: {Email} p: {Phone}){Constants.vbCrLf}{Constants.vbCrLf}{Address.Display()}{Constants.vbCrLf}";

		/// <summary>
		/// Determine if a customer object is equal to another
		/// </summary>
		/// <param name="left"></param>
		/// <param name="right"></param>
		public static bool operator ==(Customer left, Customer right)
		{
			return !(left is null ^ right is null) && left.FirstName == right.FirstName && left.LastName == right.LastName && left.Address == right.Address
			&& left.Email == right.Email && left.Phone == right.Phone && left.Joined == right.Joined;
		}

		/// <summary>
		/// Determine if a customer object is not equal to another
		/// </summary>
		/// <param name="left"></param>
		/// <param name="right"></param>
		public static bool operator !=(Customer left, Customer right) => !(left == right);

		/// <inheritdoc/>
		public override bool Equals(object obj) => this == (obj as Customer);

		/// <inheritdoc/>
		public override int GetHashCode() => base.GetHashCode();
	}
}
